### PR TITLE
Bootstrap agent tools

### PR DESCRIPTION
**Pull request by SimonRichardson**
~~Requires https://github.com/juju/juju/pull/16647 to land.~~

----

Moves the bootstrap prepareTools that was previously run by the
cmd/agent/bootstrap state command, to the bootstrap worker. The
strategy is actually in the internal package, which gives us
freedom to implement the strategy based on the dependencies. The
uploading of the tools only happens on in IAAS models. There is no
if isCAAS check, it's up to the dependency engine to give us the right
dependencies. The code path becomes branchless on that dependency.

The userdataconfig removed the tools binary after it ran, this was
problematic as the worker isn't guaranteed to have run with in that
window. Removing that code and pushing it into the strategy now ensures
it correctly cleans up it's own concern. If the upload fails, the
tools file will still be there, allowing us to retry later.

Lastly, the big major change is to run the bootstrap worker only if
the bootstrap-params file is available. Only once the bootstrap worker
has completed do we remove that file. This should allow us to perform
retries at a later time if we want to make bootstrapping more reliable.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```shj
$ juju bootstrap lxd test --build-agent
$ juju enable-ha
$ juju add-model default
$ juju deploy ubuntu
```

## Links

**Jira card:** JUJU-5136

### CODE REVIEW

File: worker/bootstrap/manifold.go [hunk 1]
```diff
@@ -140,44 +189,38 @@ func Manifold(config ManifoldConfig) dependency.Manifold {
...
+	return bootstrap.PopulateAgentBinary(ctx, dataDir, storage, logger)
 }
 
-// RequiresBootstrap returns true if the controller application does not exist.
-func RequiresBootstrap(appService ApplicationService) (bool, error) {
-	ok, err := appService.ApplicationExists(application.ControllerApplicationName)
-	if err != nil {
+// RequiresBootstrap returns true if the bootstrap params file exists.
+// It is expected at the end of bootstrap that the file is removed.
+func RequiresBootstrap(config agent.Config) (bool, error) {
```
Comment (manadart) [inline worker/bootstrap/manifold.go:213]: Wouldn't it be better to name this specifically, say `BootstrapParamsFileExists` with a comment as to what it is a proxy for?

File: worker/bootstrap/manifold.go [hunk 2]
```diff
@@ -140,44 +189,38 @@ func Manifold(config ManifoldConfig) dependency.Manifold {
...
-	if err == nil {
-		return true, nil
-	}
-	if errors.Is(err, errors.NotFound) {
-		return false, nil
-	}
-	return false, errors.Annotatef(err, "application exists")
+// CompletesBootstrap removes the bootstrap params file, completing the
+// bootstrap process.
+func CompletesBootstrap(config agent.Config) error {
```
Comment (manadart) [inline worker/bootstrap/manifold.go:223]: Same here. `RemoveBootstrapParamsFile`.

File: internal/bootstrap/agentbinary.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	logger.Debugf("Adding agent binary: %v", agentTools.Version)
+	if err := storage.Add(ctx, bytes.NewReader(data), metadata); err != nil {
+		return errors.Trace(err)
+	}
+
+	// Ensure that we remove the agent binary from disk.
+	if err := os.Remove(binaryPath); err != nil {
+		logger.Warningf("failed to remove agent binary: %v", err)
+	}
+	// Remove the sha which validates the agent binary file.
```
Comment (manadart) [inline internal/bootstrap/agentbinary.go:80]: "that validates".

File: worker/bootstrap/manifold.go [hunk 4]
```diff
@@ -75,19 +112,40 @@ func Manifold(config ManifoldConfig) dependency.Manifold {
...
+			}
+
 			var a agent.Agent
 			if err := context.Get(config.AgentName, &a); err != nil {
 				return nil, err
 			}
 
-			var bootstrapUnlocker gate.Unlocker
-			if err := context.Get(config.BootstrapGateName, &bootstrapUnlocker); err != nil {
+			// If the controller application exists, then we don't need to
```
Comment (manadart) [inline worker/bootstrap/manifold.go:132]: See here, `RequiresBootstrap` no longer checks if the app exists, so this becomes misleading.

File: worker/bootstrap/manifold.go (hunk 1 above)
Comment (SimonRichardson) [reply worker/bootstrap/manifold.go:213]: Yeah, originally we checked whether the application existed or not, so it made sense to be generalised about what we were waiting for. Now that we've focused in on the bootstrap file, I'm happy to change it.

**Pull request merged**